go-controller/pkg/clustermanager/secondary_network_cluster_manager.go

```diff
@@ -0,0 +1,145 @@
+package clustermanager
+
+import (
+	"github.com/containernetworking/cni/pkg/types"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+
+	ovncnitypes "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/cni/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
+	nad "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/network-attach-def-controller"
+	ovntypes "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+// secondaryNetworkClusterManager object manages the multi net-attach-def controllers.
+// It implements networkAttachDefController.NetworkControllerManager and can be used
+// by NetAttachDefinitionController to add and delete NADs.
+type secondaryNetworkClusterManager struct {
+	// net-attach-def controller handle net-attach-def and create/delete network controllers
+	nadController *nad.NetAttachDefinitionController
+	client        clientset.Interface
+	ovnClient     *util.OVNClusterManagerClientset
+	kube          kube.Interface
+	watchFactory  *factory.WatchFactory
+	stopChan      chan struct{}
+}
+
+func newSecondaryNetworkClusterManager(ovnClient *util.OVNClusterManagerClientset,
+	wf *factory.WatchFactory, recorder record.EventRecorder) *secondaryNetworkClusterManager {
+	klog.Infof("Creating new multi network cluster manager")
+	kube := &kube.Kube{
+		KClient: ovnClient.KubeClient,
+	}
+	sncm := &secondaryNetworkClusterManager{
+		ovnClient:    ovnClient,
+		client:       ovnClient.KubeClient,
+		kube:         kube,
+		watchFactory: wf,
+		stopChan:     make(chan struct{}),
+	}
+	sncm.nadController = nad.NewNetAttachDefinitionController(
+		"cluster-manager", sncm, ovnClient.NetworkAttchDefClient, recorder)
+	return sncm
+}
+
+// Start starts the secondary layer3 controller, handles all events and creates all needed logical entities
+func (sncm *secondaryNetworkClusterManager) Start() error {
+	klog.Infof("Starting cluster-manager/net-attach-def-controller")
+	go func() {
+		// nadController Run blocks until the stopChan is closed.
+		_ = sncm.nadController.Run(sncm.stopChan)
+	}()
+	return nil
+}
+
+func (sncm *secondaryNetworkClusterManager) Stop() {
+	close(sncm.stopChan)
```

Comment:
nadController probably never stops. Its `Run` method waits until all the workers stop with `wg.Wait()` but that could not happen until the queue is shutdown which is done in a defer of `Run` itself; so it deadlocks.

I think it would make sense for `wg.Wait()` to be done in `Stop` just after the channel is closed.